### PR TITLE
PHOENIX-6519 Make SchemaTool work with lower case table and column names

### PR DESCRIPTION
**Pull request by richardantal**

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 12s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  20m 23s |  master passed  |
| +0 |  hbaserecompile  |  32m 34s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 39s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 56s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 13s |  master passed  |
| +0 :ok: |  spotbugs  |   3m 40s |  phoenix-core in master has 964 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 47s |  phoenix-tools in master has 5 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  16m 59s |  the patch passed  |
| +0 |  hbaserecompile  |  24m 50s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 36s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 36s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 44s |  phoenix-core: The patch generated 33 new + 712 unchanged - 0 fixed = 745 total (was 712)  |
| -1 :x: |  checkstyle  |   0m 16s |  phoenix-tools: The patch generated 15 new + 213 unchanged - 12 fixed = 228 total (was 225)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   1m  6s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   4m 25s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 119m 10s |  phoenix-core in the patch passed.  |
| -1 :x: |  unit  |   3m 30s |  phoenix-tools in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 24s |  The patch does not generate ASF License warnings.  |
|  |   | 199m 48s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1275 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux 015f7b580e53 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 6842c1d |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/artifact/yetus-general-check/output/patch-unit-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/testReport/ |
| Max. process+thread count | 11543 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1275/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 1]
```diff
@@ -1238,6 +1238,78 @@ public static String getNormalizedColumnName(ColumnParseNode columnParseNode) {
...
+
+    private static boolean quotesNeededForSchema(String name) {
+        if (Strings.isNullOrEmpty(name) || name.equals(QueryConstants.DEFAULT_COLUMN_FAMILY)) {
+            return false;
+        }
+        return quotesNeededForTable(name);
+    }
+
+    private static boolean quotesNeededForColumn(String name) {
+        if (!name.equals("_INDEX_ID") && "_".equals(String.valueOf(name.charAt(0)))) {
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1274]: Why do we need the special case for tables beginning with underscore ?
Also getting the first character as string would can probably done in a less circumpect way.

**Review by stoty — COMMENTED** [review]
In general, you may want to go over the code, and make variable names more consistent.
i.e. add _Quoted_ to quoted names, etc.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java (hunk 1 above)
Comment (richardantal) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1274]: If we want to create a table named _FOO we need quotes around that. The same applies for column names.
The difference is that when we create an index, _INDEX_ID column is added automatically and we do not want to have quotes around that. (But it should be possible to create a table called "_INDEX_ID" even if it an abnormal name.)

Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1274]: Thanks for the explanation.
`name.startsWith("_")`
would be a far simpler check. (In the other methods as well.

Comment (richardantal) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1274]: Sure, I'll upload a new change soon.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 2]
```diff
@@ -1238,6 +1238,78 @@ public static String getNormalizedColumnName(ColumnParseNode columnParseNode) {
...
+
+    private static boolean quotesNeededForTable(String name) {
+        if (name.startsWith("_")) {
+            return true;
+        }
+        return isQuotesNeeded(name) || containsLowerCase(name);
+    }
+
+    public static String formatSchemaName(String name) {
+        if (quotesNeededForSchema(name)) {
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1288]: Does the leading _ issue not apply here ?

Comment (richardantal) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1288]: It does, in `quotesNeededForSchema` we check for some extra schema specific stuff but I call the `quotesNeededForTable` there.

**Review by stoty — APPROVED** [review]
+1 LGTM

**Comment by richardantal** [comment]
Thank you @stoty for the reviews!

**Pull request merged**